util/containerutil/frontend_test.go

```diff
@@ -592,23 +594,120 @@ func isBinaryInstalled(ctx context.Context, binary string) bool {
+		return fmt.Errorf("failed to pull image %s: %w", image, pullErr)
+	}
 	for _, name := range names {
-		cmd := exec.CommandContext(ctx, feBinary, "run", "-d", "--name", name, "docker.io/library/nginx:1.21", `-text="test"`)
-		output, createErr := cmd.CombinedOutput()
-		if createErr != nil {
-			// the frontend exists but is non-functional. This is... not likely to work at all.
-			err = multierror.Append(err, errors.Wrap(createErr, string(output)))
-		}
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			cmd := exec.CommandContext(ctx, feBinary, "run", "-d", "--rm", "--name", name, image, "sh", "-c", `echo output stream&&>&2 echo error stream&&sleep 100`)
+			output, createErr := cmd.CombinedOutput()
+			m.Lock()
+			defer m.Unlock()
+			if createErr != nil {
+				// the frontend exists but is non-functional. This is... not likely to work at all.
+				err = multierror.Append(err, errors.Wrap(createErr, string(output)))
+			}
+		}(name)
 	}
+	wg.Wait()
+	return err
+}
 
-	return func() {
-		for _, name := range names {
-			cmd := exec.CommandContext(ctx, feBinary, "rm", "-f", name)
-			_ = cmd.Run() // Just best effort
-		}
-	}, nil
+// pullImageIfNecessary will only pull the image if it does not exist locally
+// This helps us avoid unauthenticated rate limits in tests
+func pullImageIfNecessary(ctx context.Context, feBinary string, image string) error {
+	cmd := exec.CommandContext(ctx, feBinary, "inspect", "--type=image", image)
+	_, inspectErr := cmd.CombinedOutput()
+	if inspectErr == nil {
+		// If we are able to inspect the image then it must exist locally
+		return nil
+	}
+	cmd = exec.CommandContext(ctx, feBinary, "pull", image)
+	_, pullErr := cmd.CombinedOutput()
+	if pullErr != nil {
+		return fmt.Errorf("failed to pull image %s: %w", image, pullErr)
+	}
+	return nil
+}
+
+func removeContainers(ctx context.Context, feBinary string, names ...string) error {
+	var err error
+	m := sync.Mutex{}
+	wg := sync.WaitGroup{}
+	for _, name := range names {
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			removeCmd := exec.CommandContext(ctx, feBinary, "rm", "-f", name)
+			_, removeErr := removeCmd.CombinedOutput()
+			m.Lock()
+			defer m.Unlock()
+			if removeErr != nil {
+				err = multierror.Append(err, fmt.Errorf("failed to remove container %s", name))
+			}
+		}(name)
+	}
+	wg.Wait()
+	return err
+}
+
+func waitForContainers(ctx context.Context, feBinary string, names ...string) error {
+	var err error
+	m := sync.Mutex{}
+	wg := sync.WaitGroup{}
+	for _, name := range names {
+		const maxAttempts = 100
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			attempts := 0
+			for attempts < maxAttempts {
+				attempts++
+				// docker inspect -f {{.State.Running}} CONTAINERNAME`"=="true"
+				cmd := exec.CommandContext(ctx, feBinary, "inspect", "-f", "{{.State.Running}}", name)
+				output, inspectErr := cmd.CombinedOutput()
+				if inspectErr != nil {
+					err = multierror.Append(err, inspectErr)
```

Comment:
just noticed a scoping-bug here, `var err error` is shared between all threads in this wait group.

it should be moved to be under:

```
go func(name string) {
    defer wg.Done()
    var err error
```

This bug exists in a couple places in this PR.

Reply:
or maybe it's just missing a `m.Lock()`

Reply:
The scope was intentional, but It was missing the `m.Lock` for this appending.
From what I can tell the other places have the correct locking.